### ISSUE TITLE
Error object is empty when passed between processes

### ISSUE THREAD

**Issue by michaelnthiessen** [stryker-mutator/stryker#141]
This repo seemed the most appropriate place to file this issue, since this is where the bug is happening.

When trying to test my project using the MochaTestRunner I kept getting this error:

```
[2016-06-04 07:11:19.167] [ERROR] Stryker - One or more tests errored in the initial test run:
    [object Object]
```

The error object originates from the [mocha test run](https://github.com/stryker-mutator/stryker-mocha-runner/blob/master/src/MochaTestRunner.ts#L35) and gets passed up through the [IsolatedTestRunnerAdapterWorker](https://github.com/stryker-mutator/stryker/blob/master/src/isolated-runner/IsolatedTestRunnerAdapterWorker.ts#L89) and into the [IsolatedTestRunnerAdapter](https://github.com/stryker-mutator/stryker/blob/master/src/isolated-runner/IsolatedTestRunnerAdapter.ts#L64), which is where it disappears.

If you log the object in both places you will see that it shows up correctly in the child process, but appears as an empty object in the parent process.

Weirdly, reconstructing the error object - and then returning this new one instead - solves this issue, and it gets passed all the way up with no issues:

``` javascript
const errorObj = {
  name: error.name,
  message: error.message,
  stack: error.stack,
};
```

The reason for this is that the Error object in Javascript is behaving differently because it is a native object, and isn't actually an [Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object). The [documentation](https://nodejs.org/api/process.html#process_process_send_message_sendhandle_options_callback) for `process.send(msg)` says that it uses `JSON.stringify(message)` internally. I tried stringifying the Error object, and all you get is an empty object: `{}`.

So essentially the problem lies with the `process.send()` function only being able to handle Objects that are serializable.

The most obvious solution for this problem is to put this reconstruction in the Worker, which would solve it for all test runners.

---

Related to this is that [this line won't correctly log the object](https://github.com/stryker-mutator/stryker/blob/master/src/Stryker.ts#L155) when it does make it to the parent process, only resulting in `[object Object]`.

**Comment by simondel** [comment]
Thanks for the issue! I notice that you get a testrun with a TestResult of type `Error`, this means something went horribly wrong. *It should only occur when the TestRunner(Mocha) crashes*. Could you tell us what the error is in the [MochaTestRunner](https://github.com/stryker-mutator/stryker-mocha-runner/blob/master/src/MochaTestRunner.ts#L35)?

I would be against reconstructing the error object in the IsolatedTestRunnerAdapterWorker because the contents of the erorr object may be different between TestRunners. I would reconstruct it in the MochaTestRunner because it's responsible for generating proper responses.

The message
```
[2016-06-04 07:11:19.167] [ERROR] Stryker - One or more tests errored in the initial test run:
    [object Object]
```
that you're getting is actually correct (from a code perspective) because [the errors in a RunResult](https://github.com/stryker-mutator/stryker-api/blob/master/src/test_runner/RunResult.ts#L37) is an array of strings. So for me this is a case of 'underlying classes were being naughty'.


What do you think about this @nicojs ?

**Comment by nicojs** [comment]
I think I actually agree with @michaelnthiessen that it should be handled in the `IsolatedTestRunnerAdapterWorker`. Agreed that the errors in a `RunResult` should be a list of strings, but its still a dynamic language. For example, both `stryker` and `stryker-mocha-runner` are written in TypeScript, but that didn't prevent a wrong typed object of slipping in.

We should also refactor the errors in run result:

```typescript
errors: Array<Error|string>;

interface Error {
  name: string;
  message: string;
  stack: string;
}
```

This refactoring should be added to the [a better test runner api issue](https://github.com/stryker-mutator/stryker-api/issues/4). For this issue, i would suggest to just do a conversion in  `IsolatedTestRunnerAdapterWorker`

**Comment by michaelnthiessen** [comment]
@simondel I had a syntax error because my tests are written in es2016. My test run command is `mocha --require ignore-styles --require babel-core/register tests/ --recursive`, but I am not sure how to supply those extra arguments programmatically yet (the --require).

**Comment by michaelnthiessen** [comment]
Are there any other places where there are error objects or similar being transmitted between processes? Those areas would potentially have bugs as well.

**Comment by nicojs** [comment]
@michaelnthiessen 
> My test run command is mocha --require ignore-styles --require babel-core/register tests/ --recursive, but I am not sure how to supply those extra arguments programmatically yet (the --require).

I see your point. I think we're missing that feature from the `stryker-mocha-runner` for now. Could you please create an issue for this [there](http://github.com/stryker-mutator/stryker-mocha-runner)? If you could provide a small demo project as well, that would be awesome, we will add that to the integration test suite as well. Should be pretty straight forward to add i think.

**Comment by simondel** [comment]
@nicojs We fixed this with #165 right?

**Comment by nicojs** [comment]
Yep :)
